In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

In [13]:
file_path = r"C:\Users\knayl\Downloads\edtech_ai_operations_workbook.xlsx"

In [15]:
excel_file = pd.ExcelFile(file_path)

In [17]:
print(excel_file.sheet_names)

['README', 'data_dictionary', 'monthly_business_kpis', 'course_category_monthly', 'support_tickets', 'instructor_monthly', 'learner_level']


### Load Individual Worksheets

In [19]:
business = pd.read_excel(file_path,sheet_name = "monthly_business_kpis")
business.head(2)

,month,segment,ai_phase,ai_features_live,marketing_leads,new_signups,paid_conversions,marketing_spend_usd,cac_usd,active_users,...,renewals_completed,renewal_rate_pct,churned_users,inactive_users_30d,support_tickets,avg_ticket_resolution_hrs,support_cost_usd,content_cost_usd,instructor_cost_usd,ai_inference_cost_usd
0,2021-03-01,Upskilling,Pre-AI,0,7140.0,608.0,97.0,23258.53,240.27,8106.0,...,1285.0,72.08,412.0,1963.0,672.0,19.18,4885.04,58479.20,99213.93,0.0
1,2022-12-01,Higher Ed,Pre-AI,0,3717.0,304.0,53.0,11202.58,211.93,6820.0,...,1128.0,75.18,390.0,1620.0,579.0,19.41,4472.73,39073.58,92200.15,0.0


In [22]:
courses = pd.read_excel(file_path,sheet_name ="course_category_monthly")
courses.head(2)

,month,course_category,ai_phase,enrollments,active_courses,new_courses_launched,ai_assisted_content_pct,content_production_hours,content_cost_usd,avg_completion_rate_pct,avg_quiz_score,csat_score,refund_rate_pct,ai_tutor_queries
0,2023-07-01,Math,AI Adoption,8036.0,38,3,16.66,356.49,13708.41,62.49,60.55,3.82,2.39,4638.0
1,2025-11-01,Language Learning,Post-AI,5241.0,48,5,83.39,528.47,NaN,59.70,63.89,3.88,2.37,16310.0


In [24]:
support = pd.read_excel(file_path,sheet_name = "support_tickets")
support.head(2)

,ticket_id,created_at,learner_id,segment,issue_category,channel,handled_by,priority,first_response_mins,resolution_hrs,escalated,reopened,csat_score,cost_usd,ai_phase
0,T504296,2025-10-12 02:37,L40461,K-12,Billing/Refund,AI Chatbot,AI Bot,Medium,0.9,1.24,0,0,5.0,0.95,Post-AI
1,T502722,2023-11-15 02:05,L67896,Test Prep,Billing/Refund,In-App,Human Agent,Low,410.4,6.74,0,0,5.0,8.29,AI Adoption


In [28]:
instructors = pd.read_excel(file_path,sheet_name="instructor_monthly")
instructors.head(2)

,instructor_id,month,subject_area,employment_type,live_teaching_hrs,doubt_resolution_hrs,grading_hrs,content_units_created,content_creation_hrs,ai_tool_usage_pct,students_assigned,avg_learner_rating,instructor_cost_usd,ai_phase
0,I2000,2025-08-01,Programming,full-time,18.09,5.24,3.60,1.0,5.71,61.58,80.0,4.47,1169.23,Post-AI
1,I2031,2024-10-01,Business & Finance,Full-time,57.62,24.73,18.76,10.0,67.41,65.18,317.0,4.42,6570.53,Post-AI


In [30]:
learners = pd.read_excel(file_path,sheet_name="learner_level")
learners.head(2)

,record_id,learner_id,signup_date,country,segment,course_category,plan_type,acquisition_channel,ai_feature_used,ai_sessions_per_week,...,avg_ticket_resolution_hrs,csat_score,nps_score,monthly_fee_usd,ai_inference_cost_usd,cost_to_serve_usd,refund_flag,churned,last_active_date,tenure_months
0,R103700,L15390,2021-11-13,United States,Test Prep,Science,Basic,Paid Social,AI Quiz Generator,1.8,...,7.6,4.0,1.0,9.0,0.46,6.16,0,0,2026-08-31,58.0
1,R101762,L89375,2021-01-22,India,Test Prep,Programming,Basic,Paid Social,NaN,0.0,...,15.9,3.0,5.0,9.0,0.00,7.30,0,0,2026-08-20,67.0


### Data Quality Audit

#### Business Data

In [34]:
business.shape

(272, 34)

In [36]:
business.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 272 entries, 0 to 271
Data columns (total 34 columns):
 #   Column                            Non-Null Count  Dtype  
---  ------                            --------------  -----  
 0   month                             272 non-null    object 
 1   segment                           272 non-null    object 
 2   ai_phase                          272 non-null    object 
 3   ai_features_live                  272 non-null    int64  
 4   marketing_leads                   272 non-null    float64
 5   new_signups                       272 non-null    float64
 6   paid_conversions                  272 non-null    float64
 7   marketing_spend_usd               266 non-null    float64
 8   cac_usd                           271 non-null    float64
 9   active_users                      272 non-null    float64
 10  ai_active_users                   272 non-null    float64
 11  ai_adoption_pct                   272 non-null    float64
 12  avg_ai_s

1. Change the month data_type to Date and also name as Month_Date

In [47]:
business.loc[
    ~business["month"].str.match(r"^\d{4}-\d{2}-\d{2}$", na=False),
    "month"
]

48           02/2025
73          Jun-2026
85           2023/07
96          Nov-2022
103         Oct-2022
112          01/2021
125         Nov-2025
138          2021/06
145         Dec-2025
158         Aug-2021
201     January 2025
209       April 2025
258    November 2025
Name: month, dtype: object

In [51]:
business["month"] = pd.to_datetime(
    business["month"],
    format="mixed"
)

In [55]:
business.rename(columns={"month": "month_date"}, inplace=True)

In [59]:
business.isnull().sum()

month_date                          0
segment                             0
ai_phase                            0
ai_features_live                    0
marketing_leads                     0
new_signups                         0
paid_conversions                    0
marketing_spend_usd                 6
cac_usd                             1
active_users                        0
ai_active_users                     0
ai_adoption_pct                     0
avg_ai_sessions_per_ai_user_week    6
total_sessions                      0
total_learning_hours                6
lessons_completed                   0
avg_completion_rate_pct             6
avg_quiz_score                      0
csat_score                          6
nps_score                           6
paying_users                        0
subscription_revenue_usd            0
arpu_usd                            0
renewals_due                        0
renewals_completed                  6
renewal_rate_pct                    0
churned_user

2. Finding and Removing Duplicates

In [63]:
business.duplicated(subset=["month_date", "segment"]).sum()

3

In [65]:
business.duplicated().sum()

3

In [69]:
business[
    business.duplicated(
        subset=["month_date", "segment"],
        keep=False
    )
].sort_values(["month_date", "segment"])

,month_date,segment,ai_phase,ai_features_live,marketing_leads,new_signups,paid_conversions,marketing_spend_usd,cac_usd,active_users,...,renewals_completed,renewal_rate_pct,churned_users,inactive_users_30d,support_tickets,avg_ticket_resolution_hrs,support_cost_usd,content_cost_usd,instructor_cost_usd,ai_inference_cost_usd
49,2022-03-01,Test Prep,Pre-AI,0,7653.0,686.0,117.0,24938.37,214.02,10214.0,...,1572.0,69.97,552.0,2328.0,879.0,20.31,6256.19,48742.45,99579.50,0.00
199,2022-03-01,Test Prep,Pre-AI,0,7653.0,686.0,117.0,24938.37,214.02,10214.0,...,1572.0,69.97,552.0,2328.0,879.0,20.31,6256.19,48742.45,99579.50,0.00
76,2022-12-01,Upskilling,Pre-AI,0,5813.0,442.0,81.0,17033.55,209.24,8153.0,...,1284.0,71.59,410.0,1902.0,755.0,19.48,5701.21,50447.92,100760.38,0.00
113,2022-12-01,Upskilling,Pre-AI,0,5813.0,442.0,81.0,17033.55,209.24,8153.0,...,1284.0,71.59,410.0,1902.0,755.0,19.48,5701.21,50447.92,100760.38,0.00
29,2026-03-01,Test Prep,Post-AI,5,6886.0,649.0,122.0,26205.83,214.62,12086.0,...,2207.0,75.68,536.0,2471.0,833.0,13.75,4266.66,41503.51,109921.40,6823.48
213,2026-03-01,Test Prep,Post-AI,5,6886.0,649.0,122.0,26205.83,214.62,12086.0,...,2207.0,75.68,536.0,2471.0,833.0,13.75,4266.66,41503.51,109921.40,6823.48


In [71]:
business = business.drop_duplicates()

In [73]:
business.duplicated().sum()

0

3. Range / validity checks

In [77]:
business.describe()

,month_date,ai_features_live,marketing_leads,new_signups,paid_conversions,marketing_spend_usd,cac_usd,active_users,ai_active_users,ai_adoption_pct,...,renewals_completed,renewal_rate_pct,churned_users,inactive_users_30d,support_tickets,avg_ticket_resolution_hrs,support_cost_usd,content_cost_usd,instructor_cost_usd,ai_inference_cost_usd
count,269,269.000000,269.000000,269.000000,269.000000,263.000000,268.000000,269.000000,269.000000,269.000000,...,263.000000,269.000000,269.000000,269.000000,263.000000,269.000000,269.000000,263.000000,269.000000,269.000000
mean,2023-10-17 15:52:51.747211776,2.315985,6339.858736,591.081784,110.204461,20922.511521,192.352015,10518.665428,3183.048327,28.644572,...,1852.041825,75.788104,541.208178,2303.438662,4590.813688,15.882045,5136.325911,47574.673308,105028.767249,3664.051784
min,2021-01-01 00:00:00,0.000000,3173.000000,279.000000,48.000000,9898.390000,0.000000,6524.000000,0.000000,-5.000000,...,1042.000000,69.190000,329.000000,1377.000000,-40.000000,9.540000,2435.140000,30510.970000,85661.500000,0.000000
25%,2022-05-01 00:00:00,0.000000,4841.000000,451.000000,82.000000,16076.000000,168.815000,8177.000000,0.000000,0.000000,...,1357.000000,73.220000,410.000000,1747.000000,603.500000,12.920000,3827.220000,41203.365000,95937.720000,0.000000
50%,2023-10-01 00:00:00,2.000000,5797.000000,550.000000,104.000000,19640.060000,192.115000,9823.000000,1628.000000,19.400000,...,1676.000000,75.270000,509.000000,2129.000000,736.000000,16.640000,4796.470000,46306.210000,102882.690000,2263.670000
75%,2025-04-01 00:00:00,5.000000,7459.000000,686.000000,130.000000,24560.875000,213.327500,12309.000000,6148.000000,56.430000,...,2272.500000,78.010000,650.000000,2613.000000,962.500000,18.800000,6057.830000,54765.815000,113303.210000,6785.560000
max,2026-08-01 00:00:00,5.000000,12275.000000,1096.000000,241.000000,39329.700000,284.610000,17899.000000,11426.000000,137.000000,...,3540.000000,83.460000,952.000000,3687.000000,999999.000000,21.010000,9783.730000,66915.590000,132617.550000,12897.210000
std,NaN,2.168414,1934.356342,179.379384,35.931618,6445.588285,30.906467,3079.552410,3434.714557,29.014526,...,605.528621,3.290200,155.679880,662.239073,61614.227637,3.179881,1801.707683,8670.925174,11350.695299,3763.182534


In [79]:
business["renewal_rate_pct"].describe()

count    269.000000
mean      75.788104
std        3.290200
min       69.190000
25%       73.220000
50%       75.270000
75%       78.010000
max       83.460000
Name: renewal_rate_pct, dtype: float64

In [81]:
business["avg_completion_rate_pct"].describe()

count    263.000000
mean      49.646426
std        5.510039
min       -3.000000
25%       47.345000
50%       49.300000
75%       52.215000
max      104.200000
Name: avg_completion_rate_pct, dtype: float64

In [85]:
business[
    (business["avg_completion_rate_pct"] < 0) |
    (business["avg_completion_rate_pct"] > 100)
][["month_date", "segment", "avg_completion_rate_pct"]]

,month_date,segment,avg_completion_rate_pct
133,2026-06-01,Upskilling,-3.0
251,2026-06-01,Higher Ed,104.2


completion_rate should be between 0 - 100 %

In [89]:
business.loc[
    (business["avg_completion_rate_pct"] < 0) |
    (business["avg_completion_rate_pct"] > 100),
    "avg_completion_rate_pct"
] = pd.NA

In [91]:
business["avg_completion_rate_pct"].describe()

count    261.000000
mean      49.639119
std        2.913302
min       43.710000
25%       47.370000
50%       49.300000
75%       52.210000
max       57.300000
Name: avg_completion_rate_pct, dtype: float64

In [94]:
business["ai_adoption_pct"].describe()

count    269.000000
mean      28.644572
std       29.014526
min       -5.000000
25%        0.000000
50%       19.400000
75%       56.430000
max      137.000000
Name: ai_adoption_pct, dtype: float64

In [96]:
business[
    (business["ai_adoption_pct"] < 0) |
    (business["ai_adoption_pct"] > 100)
][["month_date", "segment", "ai_adoption_pct"]]

,month_date,segment,ai_adoption_pct
111,2021-04-01,Test Prep,-5.0
156,2024-05-01,K-12,137.0


ai_adoption_rate should be between 0 - 100 %

In [98]:
business.loc[
    (business["ai_adoption_pct"] < 0) |
    (business["ai_adoption_pct"] > 100),
    "ai_adoption_pct"
] = pd.NA

In [100]:
business["ai_adoption_pct"].describe()

count    267.000000
mean      28.364757
std       28.278960
min        0.000000
25%        0.000000
50%       19.400000
75%       56.290000
max       78.020000
Name: ai_adoption_pct, dtype: float64

In [105]:
business["nps_score"].describe()

count    263.000000
mean      24.089163
std        3.494578
min       17.430000
25%       21.020000
50%       23.390000
75%       27.105000
max       31.850000
Name: nps_score, dtype: float64

In [107]:
business["csat_score"].describe()

count    263.000000
mean       4.235894
std        4.415812
min        0.000000
25%        3.715000
50%        3.800000
75%        3.920000
max       45.000000
Name: csat_score, dtype: float64

4. Business logic checks

In [109]:
### AI active users shouldn't normally exceed active users
business[business["ai_active_users"] > business["active_users"]]

,month_date,segment,ai_phase,ai_features_live,marketing_leads,new_signups,paid_conversions,marketing_spend_usd,cac_usd,active_users,...,renewals_completed,renewal_rate_pct,churned_users,inactive_users_30d,support_tickets,avg_ticket_resolution_hrs,support_cost_usd,content_cost_usd,instructor_cost_usd,ai_inference_cost_usd


In [111]:
### Paying users shouldn't normally exceed active users
business[business["paying_users"] > business["active_users"]]

,month_date,segment,ai_phase,ai_features_live,marketing_leads,new_signups,paid_conversions,marketing_spend_usd,cac_usd,active_users,...,renewals_completed,renewal_rate_pct,churned_users,inactive_users_30d,support_tickets,avg_ticket_resolution_hrs,support_cost_usd,content_cost_usd,instructor_cost_usd,ai_inference_cost_usd


In [115]:
### Renewals completed shouldn't exceed renewals due
business[business["renewals_completed"] > business["renewals_due"]]

,month_date,segment,ai_phase,ai_features_live,marketing_leads,new_signups,paid_conversions,marketing_spend_usd,cac_usd,active_users,...,renewals_completed,renewal_rate_pct,churned_users,inactive_users_30d,support_tickets,avg_ticket_resolution_hrs,support_cost_usd,content_cost_usd,instructor_cost_usd,ai_inference_cost_usd


In [117]:
### Churned users shouldn't be negative
business[business["churned_users"] < 0]

,month_date,segment,ai_phase,ai_features_live,marketing_leads,new_signups,paid_conversions,marketing_spend_usd,cac_usd,active_users,...,renewals_completed,renewal_rate_pct,churned_users,inactive_users_30d,support_tickets,avg_ticket_resolution_hrs,support_cost_usd,content_cost_usd,instructor_cost_usd,ai_inference_cost_usd


In [119]:
### Costs and revenue shouldn't be negative
business[
    (business["subscription_revenue_usd"] < 0) |
    (business["support_cost_usd"] < 0) |
    (business["content_cost_usd"] < 0) |
    (business["instructor_cost_usd"] < 0) |
    (business["ai_inference_cost_usd"] < 0)
]

,month_date,segment,ai_phase,ai_features_live,marketing_leads,new_signups,paid_conversions,marketing_spend_usd,cac_usd,active_users,...,renewals_completed,renewal_rate_pct,churned_users,inactive_users_30d,support_tickets,avg_ticket_resolution_hrs,support_cost_usd,content_cost_usd,instructor_cost_usd,ai_inference_cost_usd


5. Categorical consistency

In [125]:
business["segment"].value_counts()

segment
Higher Ed     66
Upskilling    65
Test Prep     64
K-12          64
upskilling     3
test prep      2
K 12           2
k-12           1
higher ed      1
HIGHER ED      1
Name: count, dtype: int64

In [129]:
### standardize inconsistent category naming
business["segment"] = (
    business["segment"]
    .str.strip()
    .str.lower()
    .replace({
        "higher ed": "Higher Ed",
        "upskilling": "Upskilling",
        "test prep": "Test Prep",
        "k 12": "K-12",
        "k-12": "K-12"
    })
)

In [133]:
business["segment"].value_counts()

segment
Upskilling    68
Higher Ed     68
K-12          67
Test Prep     66
Name: count, dtype: int64

In [127]:
business["ai_phase"].value_counts()

ai_phase
Post-AI        116
Pre-AI          99
AI Adoption     44
Adoption         3
pre-ai           2
Post AI          2
Pre AI           2
post-AI          1
Name: count, dtype: int64

In [141]:
### standardize inconsistent category naming
business["ai_phase"] = business["ai_phase"].replace({
    "pre-ai": "Pre-AI",
    "Pre AI": "Pre-AI",
    "Adoption": "AI Adoption",
    "Post AI": "Post-AI",
    "post-AI": "Post-AI"
})

In [143]:
business["ai_phase"].value_counts()

ai_phase
Post-AI        119
Pre-AI         103
AI Adoption     47
Name: count, dtype: int64

#### Courses Table

In [149]:
courses.shape

(480, 14)

In [151]:
courses.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 480 entries, 0 to 479
Data columns (total 14 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   month                     480 non-null    object 
 1   course_category           480 non-null    object 
 2   ai_phase                  480 non-null    object 
 3   enrollments               480 non-null    float64
 4   active_courses            480 non-null    int64  
 5   new_courses_launched      480 non-null    int64  
 6   ai_assisted_content_pct   480 non-null    float64
 7   content_production_hours  480 non-null    float64
 8   content_cost_usd          466 non-null    float64
 9   avg_completion_rate_pct   480 non-null    float64
 10  avg_quiz_score            466 non-null    float64
 11  csat_score                466 non-null    float64
 12  refund_rate_pct           466 non-null    float64
 13  ai_tutor_queries          466 non-null    float64
dtypes: float64

1.Changing Month Datatype to Date

In [153]:
courses["month"].unique()

array(['2023-07-01', '2025-11-01', '2022-11-01', '2026-07-01',
       '2024-09-01', '2025-06-01', '12/2025', '2023-05-01', '2025-01-01',
       '2023-04-01', '2025-12-01', '2022-08-01', '2021-02-01',
       '2024-07-01', '2023-03-01', '2023-02-01', '2025-09-01',
       '2021-09-01', '2021-04-01', '2026-05-01', '2022-02-01',
       '2021-07-01', 'May 2026', '2023-10-01', '2022-01-01', '2021-11-01',
       '2021-08-01', '2025-05-01', '2026-01-01', '2024-06-01',
       '2021-05-01', '2026/02', '2025-10-01', '2021-06-01', '2023-08-01',
       '2023-09-01', '2022-06-01', '2024-10-01', '2023-11-01',
       '2024-01-01', '2021-01-01', '2024-04-01', '2021-03-01',
       '2024-02-01', 'August 2025', '2024-11-01', '2024/12', '2022-04-01',
       '2021-12-01', '2024-05-01', '2024-08-01', '2022-12-01',
       '2025-07-01', '2022-09-01', '2022-05-01', '2022-03-01',
       '2026-08-01', '2023-12-01', '2022-07-01', '2025-02-01',
       '2024-12-01', '2026-02-01', '2026-04-01', '03/2023', 'Jan-2022',


In [155]:
courses["month"] = pd.to_datetime(
    courses["month"],
    format="mixed"
)

In [157]:
courses["month"].dtype

dtype('<M8[ns]')

In [160]:
courses.isnull().sum()

month                        0
course_category              0
ai_phase                     0
enrollments                  0
active_courses               0
new_courses_launched         0
ai_assisted_content_pct      0
content_production_hours     0
content_cost_usd            14
avg_completion_rate_pct      0
avg_quiz_score              14
csat_score                  14
refund_rate_pct             14
ai_tutor_queries            14
dtype: int64

2. Checking for duplicates

In [162]:
courses.duplicated().sum()

4

In [166]:
courses.duplicated(subset=["month", "course_category"]).sum()

4

In [168]:
courses[courses.duplicated(
    subset=["month", "course_category"],
    keep=False
)].sort_values(["month", "course_category"])

,month,course_category,ai_phase,enrollments,active_courses,new_courses_launched,ai_assisted_content_pct,content_production_hours,content_cost_usd,avg_completion_rate_pct,avg_quiz_score,csat_score,refund_rate_pct,ai_tutor_queries
158,2021-04-01,Programming,Post AI,7313.0,29,4,0.0,597.59,23621.80,52.12,59.49,3.87,2.86,0.0
241,2021-04-01,Programming,Post AI,7313.0,29,4,0.0,597.59,23621.80,52.12,59.49,3.87,2.86,0.0
192,2022-03-01,Exam Prep (JEE/NEET/SAT),Pre-AI,8007.0,32,2,0.0,280.96,10214.74,50.72,58.12,3.90,2.81,0.0
193,2022-03-01,Exam Prep (JEE/NEET/SAT),Pre-AI,8007.0,32,2,0.0,280.96,10214.74,50.72,58.12,3.90,2.81,0.0
286,2022-11-01,Language Learning,Pre-AI,4936.0,35,3,0.0,413.51,16072.96,56.57,60.86,3.70,2.61,0.0
294,2022-11-01,Language Learning,Pre-AI,4936.0,35,3,0.0,413.51,16072.96,56.57,60.86,3.70,2.61,0.0
104,2023-11-01,Exam Prep,AI Adoption,3344.0,45,4,26.8,499.74,18709.30,-8.00,60.37,3.88,2.59,3633.0
150,2023-11-01,Exam Prep,AI Adoption,3344.0,45,4,26.8,499.74,18709.30,-8.00,60.37,3.88,2.59,3633.0


In [170]:
courses = courses.drop_duplicates()

In [172]:
courses.duplicated().sum()

0

3.Range / validity checks

In [176]:
courses.describe()

,month,enrollments,active_courses,new_courses_launched,ai_assisted_content_pct,content_production_hours,content_cost_usd,avg_completion_rate_pct,avg_quiz_score,csat_score,refund_rate_pct,ai_tutor_queries
count,476,476.000000,476.000000,476.000000,476.000000,476.000000,462.000000,476.000000,462.000000,462.000000,462.000000,462.000000
mean,2023-10-16 14:07:03.529411840,6131.897059,40.121849,3.621849,34.392647,440.365315,16689.681364,56.988298,61.007381,3.876342,2.463636,11031.495671
min,2021-01-01 00:00:00,-120.000000,28.000000,0.000000,-10.000000,0.000000,0.000000,-8.000000,57.200000,3.670000,1.780000,0.000000
25%,2022-05-24 06:00:00,4878.500000,34.000000,2.000000,0.000000,274.872500,10472.715000,52.020000,59.972500,3.800000,2.290000,0.000000
50%,2023-10-16 12:00:00,6399.000000,40.000000,3.000000,23.470000,419.660000,16010.005000,57.220000,60.910000,3.870000,2.480000,5563.500000
75%,2025-03-08 18:00:00,7398.250000,46.000000,5.000000,72.922500,583.630000,22482.420000,61.217500,62.052500,3.950000,2.630000,20013.500000
max,2026-08-01 00:00:00,10070.000000,57.000000,12.000000,160.000000,1204.900000,42681.100000,125.000000,64.620000,4.190000,3.120000,49785.000000
std,NaN,1809.811327,7.191067,1.922681,34.675971,223.148950,8478.812908,8.358160,1.442977,0.098792,0.248315,12329.871310


In [178]:
courses.columns

Index(['month', 'course_category', 'ai_phase', 'enrollments', 'active_courses',
       'new_courses_launched', 'ai_assisted_content_pct',
       'content_production_hours', 'content_cost_usd',
       'avg_completion_rate_pct', 'avg_quiz_score', 'csat_score',
       'refund_rate_pct', 'ai_tutor_queries'],
      dtype='object')

In [180]:
courses["ai_assisted_content_pct"].describe()

count    476.000000
mean      34.392647
std       34.675971
min      -10.000000
25%        0.000000
50%       23.470000
75%       72.922500
max      160.000000
Name: ai_assisted_content_pct, dtype: float64

In [188]:
courses[
    (courses["ai_assisted_content_pct"] < 0) |
    (courses["ai_assisted_content_pct"] > 100)
]

,month,course_category,ai_phase,enrollments,active_courses,new_courses_launched,ai_assisted_content_pct,content_production_hours,content_cost_usd,avg_completion_rate_pct,avg_quiz_score,csat_score,refund_rate_pct,ai_tutor_queries
48,2024-10-01,Business & Finance,Post-AI,4637.0,43,2,-10.0,213.32,8082.06,59.62,64.62,3.85,2.61,20347.0
295,2023-06-01,Programming,AI Adoption,0.0,38,4,-10.0,473.83,18595.29,51.53,60.28,3.79,2.65,3544.0
302,2024-07-01,Language Learning,Post-AI,5261.0,42,2,-10.0,219.45,8590.99,61.87,61.09,3.96,2.50,12300.0
320,2024-10-01,Math,Post-AI,7875.0,43,1,160.0,113.00,4118.21,65.92,61.30,3.89,2.21,30405.0


In [190]:
courses.loc[
    (courses["ai_assisted_content_pct"] < 0) |
    (courses["ai_assisted_content_pct"] > 100),
    "ai_assisted_content_pct"
] = np.nan

In [182]:
courses["avg_completion_rate_pct"].describe()

count    476.000000
mean      56.988298
std        8.358160
min       -8.000000
25%       52.020000
50%       57.220000
75%       61.217500
max      125.000000
Name: avg_completion_rate_pct, dtype: float64

In [194]:
courses[
    (courses["avg_completion_rate_pct"] < 0) |
    (courses["avg_completion_rate_pct"] > 100)
]

,month,course_category,ai_phase,enrollments,active_courses,new_courses_launched,ai_assisted_content_pct,content_production_hours,content_cost_usd,avg_completion_rate_pct,avg_quiz_score,csat_score,refund_rate_pct,ai_tutor_queries
104,2023-11-01,Exam Prep,AI Adoption,3344.0,45,4,26.80,499.74,18709.30,-8.0,60.37,3.88,2.59,3633.0
106,2021-05-01,Math,Pre-AI,8832.0,29,4,0.00,643.25,24807.62,125.0,59.90,3.81,2.53,0.0
117,2024-12-01,Science,Post-AI,6396.0,44,5,72.92,489.19,18641.59,125.0,60.44,4.04,2.61,30282.0
242,2021-09-01,Data & AI,post-AI,3312.0,36,5,0.00,730.69,27881.37,125.0,57.71,3.86,2.49,0.0


In [198]:
courses.loc[
    (courses["avg_completion_rate_pct"] < 0) |
    (courses["avg_completion_rate_pct"] > 100),
    "avg_completion_rate_pct"
] = np.nan

In [200]:
courses["avg_completion_rate_pct"].describe()

count    472.000000
mean      56.693708
std        5.651172
min       43.280000
25%       52.020000
50%       57.175000
75%       61.182500
max       67.940000
Name: avg_completion_rate_pct, dtype: float64

4. Check categorical values

In [205]:
courses["course_category"].value_counts()

course_category
Language Learning           66
Exam Prep (JEE/NEET/SAT)    66
Math                        65
Business & Finance          65
Programming                 64
Data & AI                   64
Science                     63
Language                     5
Data and AI                  4
Exam Prep                    4
Programing                   4
data & ai                    2
exam prep (jee/neet/sat)     1
business & finance           1
math                         1
language learning            1
Name: count, dtype: int64

In [209]:
courses[courses["course_category"].isin(["Language", "Exam Prep"])]

,month,course_category,ai_phase,enrollments,active_courses,new_courses_launched,ai_assisted_content_pct,content_production_hours,content_cost_usd,avg_completion_rate_pct,avg_quiz_score,csat_score,refund_rate_pct,ai_tutor_queries
90,2025-05-01,Language,Pre AI,7173.0,46,5,69.64,578.27,21792.42,59.03,60.80,3.95,2.02,24939.0
104,2023-11-01,Exam Prep,AI Adoption,3344.0,45,4,26.80,499.74,18709.30,NaN,60.37,3.88,2.59,3633.0
152,2023-07-01,Language,AI Adoption,7074.0,38,6,14.89,862.71,33297.17,61.26,60.29,3.84,2.53,2983.0
221,2023-05-01,Language,AI Adoption,6487.0,37,5,10.70,589.18,22207.67,62.42,59.79,3.74,2.57,4010.0
267,2022-02-01,Exam Prep,Pre-AI,4033.0,32,7,0.00,916.79,34338.10,54.77,61.35,3.76,2.86,0.0
355,2022-12-01,Language,Pre-AI,5731.0,36,3,0.00,436.71,16606.95,60.75,58.35,3.74,2.71,0.0
399,2024-12-01,Exam Prep,Post-AI,4669.0,44,6,65.09,706.22,NaN,62.46,61.68,3.96,2.13,13292.0
445,2023-02-01,Language,Pre-AI,2297.0,42,4,0.00,518.35,19088.54,46.68,NaN,3.75,2.73,0.0
453,2025-02-01,Exam Prep,Post-AI,8595.0,45,5,73.70,507.49,18678.16,55.41,61.96,3.92,2.63,NaN


In [211]:
courses["course_category"] = courses["course_category"].replace({
    "Language": "Language Learning",
    "language learning": "Language Learning",

    "Exam Prep": "Exam Prep (JEE/NEET/SAT)",
    "exam prep (jee/neet/sat)": "Exam Prep (JEE/NEET/SAT)",

    "data & ai": "Data & AI",
    "Data and AI": "Data & AI",

    "Programing": "Programming",

    "business & finance": "Business & Finance",

    "math": "Math"
})

In [215]:
courses["course_category"].value_counts()

course_category
Language Learning           72
Exam Prep (JEE/NEET/SAT)    71
Data & AI                   70
Programming                 68
Math                        66
Business & Finance          66
Science                     63
Name: count, dtype: int64

In [207]:
courses["ai_phase"].value_counts()

ai_phase
Post-AI        201
Pre-AI         173
AI Adoption     83
post-AI          5
Post AI          5
Pre AI           4
pre-ai           3
Adoption         2
Name: count, dtype: int64

In [217]:
courses["ai_phase"] = courses["ai_phase"].replace({
    "post-AI": "Post-AI",
    "Post AI": "Post-AI",

    "Pre AI": "Pre-AI",
    "pre-ai": "Pre-AI",

    "Adoption": "AI Adoption"
})

In [219]:
courses["ai_phase"].value_counts()

ai_phase
Post-AI        211
Pre-AI         180
AI Adoption     85
Name: count, dtype: int64

5. Check negative values

In [226]:
non_negative_cols = [
    "enrollments",
    "active_courses",
    "new_courses_launched",
    "ai_assisted_content_pct",
    "content_production_hours",
    "content_cost_usd",
    "avg_completion_rate_pct",
    "avg_quiz_score",
    "csat_score",
    "refund_rate_pct",
    "ai_tutor_queries"
]

for col in non_negative_cols:
    print(col, (courses[col] < 0).sum())

enrollments 2
active_courses 0
new_courses_launched 0
ai_assisted_content_pct 0
content_production_hours 0
content_cost_usd 0
avg_completion_rate_pct 0
avg_quiz_score 0
csat_score 0
refund_rate_pct 0
ai_tutor_queries 0


In [228]:
courses[courses["enrollments"] < 0]

,month,course_category,ai_phase,enrollments,active_courses,new_courses_launched,ai_assisted_content_pct,content_production_hours,content_cost_usd,avg_completion_rate_pct,avg_quiz_score,csat_score,refund_rate_pct,ai_tutor_queries
177,2025-03-01,Math,Post-AI,-120.0,45,4,69.35,404.67,15409.87,65.37,61.21,4.01,2.69,23355.0
429,2025-09-01,Data & AI,Post-AI,-120.0,53,4,78.44,385.50,NaN,50.93,62.69,4.08,1.86,24411.0


In [230]:
courses.loc[courses["enrollments"] < 0, "enrollments"] = np.nan

In [232]:
courses[courses["enrollments"] < 0]

,month,course_category,ai_phase,enrollments,active_courses,new_courses_launched,ai_assisted_content_pct,content_production_hours,content_cost_usd,avg_completion_rate_pct,avg_quiz_score,csat_score,refund_rate_pct,ai_tutor_queries


In [234]:
(courses == 0).sum()

month                         0
course_category               0
ai_phase                      0
enrollments                   2
active_courses                0
new_courses_launched         15
ai_assisted_content_pct     182
content_production_hours     15
content_cost_usd             14
avg_completion_rate_pct       0
avg_quiz_score                0
csat_score                    0
refund_rate_pct               0
ai_tutor_queries            177
dtype: int64

In [240]:
courses[courses["content_cost_usd"] == 0]

,month,course_category,ai_phase,enrollments,active_courses,new_courses_launched,ai_assisted_content_pct,content_production_hours,content_cost_usd,avg_completion_rate_pct,avg_quiz_score,csat_score,refund_rate_pct,ai_tutor_queries
4,2024-09-01,Language Learning,Post-AI,5582.0,43,0,55.10,0.0,0.0,59.34,61.10,3.97,2.25,22821.0
37,2021-05-01,Data & AI,Pre-AI,3035.0,35,0,0.00,0.0,0.0,45.70,NaN,3.82,2.70,0.0
79,2023-04-01,Business & Finance,AI Adoption,4011.0,37,0,8.62,0.0,0.0,55.15,60.30,3.94,2.33,2081.0
80,2022-05-01,Math,Pre-AI,7398.0,33,0,0.00,0.0,0.0,61.10,NaN,3.88,3.04,0.0
146,2024-10-01,Science,Post-AI,7265.0,43,0,66.72,0.0,0.0,64.23,61.54,3.91,2.47,20669.0
190,2022-06-01,Language Learning,Pre-AI,6015.0,33,0,0.00,0.0,0.0,57.28,61.43,3.81,2.64,0.0
243,2021-03-01,Math,Pre-AI,8548.0,28,0,0.00,0.0,0.0,60.03,59.99,3.81,2.63,0.0
290,2023-02-01,Math,Pre-AI,6688.0,36,0,0.00,0.0,0.0,61.04,60.63,3.80,2.45,0.0
316,2025-07-01,Language Learning,Post-AI,5572.0,46,0,74.15,0.0,0.0,63.07,62.98,4.04,2.25,17205.0
353,2023-01-01,Language Learning,Pre-AI,4364.0,36,0,0.00,0.0,0.0,58.24,59.83,3.79,2.41,0.0


#### Support Data

In [252]:
support.shape

(5100, 15)

In [254]:
support.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5100 entries, 0 to 5099
Data columns (total 15 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   ticket_id            5100 non-null   object 
 1   created_at           5100 non-null   object 
 2   learner_id           5100 non-null   object 
 3   segment              5100 non-null   object 
 4   issue_category       5100 non-null   object 
 5   channel              4998 non-null   object 
 6   handled_by           5006 non-null   object 
 7   priority             4999 non-null   object 
 8   first_response_mins  5001 non-null   float64
 9   resolution_hrs       5000 non-null   float64
 10  escalated            5100 non-null   int64  
 11  reopened             5100 non-null   int64  
 12  csat_score           3531 non-null   float64
 13  cost_usd             5100 non-null   float64
 14  ai_phase             5100 non-null   object 
dtypes: float64(4), int64(2), object(9)
mem

Change the created_at datat type to Date

In [303]:
support["created_at"].unique()

array(['2025-10-12 02:37', '2023-11-15 02:05', '2022-01-02 16:41', ...,
       '2023-01-16 16:33', '2021-03-16 06:33', '2023-12-07 03:56'],
      dtype=object)

In [311]:
support["created_at"] = pd.to_datetime(support["created_at"], errors="coerce")

In [315]:
support["created_at"].dtype

dtype('<M8[ns]')

In [258]:
support.isnull().sum()

ticket_id                 0
created_at                0
learner_id                0
segment                   0
issue_category            0
channel                 102
handled_by               94
priority                101
first_response_mins      99
resolution_hrs          100
escalated                 0
reopened                  0
csat_score             1569
cost_usd                  0
ai_phase                  0
dtype: int64

In [260]:
support.groupby("ai_phase")["csat_score"].apply(
    lambda x: x.isna().sum()
)

ai_phase
AI Adoption    281
Post-AI        639
Pre-AI         649
Name: csat_score, dtype: int64

In [262]:
support.groupby("ai_phase")["csat_score"].apply(
    lambda x: x.isna().mean() * 100
)

ai_phase
AI Adoption    33.254438
Post-AI        30.515759
Pre-AI         30.032392
Name: csat_score, dtype: float64

CSAT has 1,569 missing values (30.76%). Missingness is relatively consistent across AI phases, ranging from 30.03% to 33.25%. Since a missing CSAT may indicate that a learner did not provide a rating, the values will be retained as missing rather than imputed. CSAT-based analysis will use only tickets with available CSAT scores.

In [265]:
support.groupby("escalated")["csat_score"].apply(
    lambda x: x.isna().mean() * 100
)

escalated
0    31.029652
1    29.375765
Name: csat_score, dtype: float64

In [267]:
support.groupby("reopened")["csat_score"].apply(
    lambda x: x.isna().mean() * 100
)

reopened
0    30.679106
1    31.568228
Name: csat_score, dtype: float64

In [269]:
support.groupby("ai_phase")[[
    "channel",
    "handled_by",
    "priority",
    "first_response_mins",
    "resolution_hrs"
]].apply(lambda x: x.isna().mean() * 100)

,channel,handled_by,priority,first_response_mins,resolution_hrs
ai_phase,,,,,
AI Adoption,2.248521,1.420118,2.011834,1.656805,1.775148
Post-AI,1.766953,1.575931,2.340019,2.196753,2.435530
Pre-AI,2.128644,2.267469,1.619621,1.804720,1.573346


Missing values were identified in six columns. CSAT had the highest missingness at 30.76%, while other affected fields had approximately 2% missing values. Missingness was compared across AI phases, escalation status, and reopening status. No substantial concentration of missing values was observed in any AI phase or ticket outcome group. Missing values will therefore be retained rather than blindly imputed or deleted.

In [272]:
support["ticket_id"].duplicated().sum()

100

In [274]:
support.duplicated().sum()

100

In [276]:
support[support["ticket_id"].duplicated(keep=False)].sort_values("ticket_id")

,ticket_id,created_at,learner_id,segment,issue_category,channel,handled_by,priority,first_response_mins,resolution_hrs,escalated,reopened,csat_score,cost_usd,ai_phase
3481,T500027,2021-01-11 20:46,L31329,Higher Ed,Technical Bug,Live Chat,Human Agent,Medium,334.5,16.42,0,0,3.0,7.95,Pre-AI
4066,T500027,2021-01-11 20:46,L31329,Higher Ed,Technical Bug,Live Chat,Human Agent,Medium,334.5,16.42,0,0,3.0,7.95,Pre-AI
3388,T500033,2021-01-13 01:09,L94949,K-12,Course Content Query,Live Chat,Human Agent,NaN,210.3,11.89,0,0,NaN,7.37,Pre-AI
3394,T500033,2021-01-13 01:09,L94949,K-12,Course Content Query,Live Chat,Human Agent,NaN,210.3,11.89,0,0,NaN,7.37,Pre-AI
23,T500111,2021-02-11 03:45,L73638,Upskilling,Assignment/Grading,Email,Human Agent,Medium,90.1,NaN,0,0,4.0,9.68,Pre-AI
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
540,T504941,2026-08-05 02:22,L90838,K-12,Technical Bug,In-App,Human Agent,Low,164.8,10.04,0,0,3.0,7.73,Post-AI
5003,T504944,2026-08-06 02:36,L72536,Test Prep,Login/Access,AI Chatbot,AI Bot,Critical,1.2,0.53,0,0,5.0,0.43,Post-AI
4740,T504944,2026-08-06 02:36,L72536,Test Prep,Login/Access,AI Chatbot,AI Bot,Critical,1.2,0.53,0,0,5.0,0.43,Post-AI
4552,T504969,2026-08-17 18:45,L95369,Higher Ed,Login/Access,AI Chatbot,AI Bot -> Human,High,3.6,7.58,1,0,4.0,4.21,Post-AI


In [278]:
support["ticket_id"].value_counts().value_counts()

count
1    4900
2     100
Name: count, dtype: int64

In [280]:
support = support.drop_duplicates()

In [282]:
support.shape

(5000, 15)

In [284]:
support["ticket_id"].duplicated().sum()

0

Duplicate check identified 100 exact duplicate records. These represented 100 ticket IDs appearing twice, with identical values across all 15 columns. One copy of each duplicate record was removed, reducing the dataset from 5,100 to 5,000 unique ticket records.

Check categorical values

In [290]:
for col in ["segment", "issue_category", "channel",
            "handled_by", "priority", "ai_phase"]:
    print("\n", col)
    print(support[col].value_counts(dropna=False))


 segment
segment
K-12          1542
Test Prep     1141
Higher Ed     1061
Upskilling    1056
test prep       49
K12             47
k12             38
higher ed       35
upskilling      31
Name: count, dtype: int64

 issue_category
issue_category
Course Content Query    1082
Billing/Refund           926
Assignment/Grading       854
Login/Access             819
Technical Bug            789
Certificate              280
login                     42
billing/refund            40
Billing                   40
Course content query      36
Tech Bug                  33
Assignment / Grading      32
Certificte                27
Name: count, dtype: int64

 channel
channel
Email         1465
AI Chatbot    1338
Live Chat     1016
Phone          721
In-App         360
NaN            100
Name: count, dtype: int64

 handled_by
handled_by
Human Agent        3420
AI Bot              914
AI Bot -> Human     373
NaN                  93
ai bot               48
Human                47
human agent          40


In [294]:
# 1. Standardize segment
segment_map = {
    "K12": "K-12",
    "k12": "K-12",
    "test prep": "Test Prep",
    "higher ed": "Higher Ed",
    "upskilling": "Upskilling"
}

support["segment"] = support["segment"].replace(segment_map)


# 2. Standardize issue_category
issue_category_map = {
    "Course content query": "Course Content Query",
    "billing/refund": "Billing/Refund",
    "Billing": "Billing/Refund",
    "Assignment / Grading": "Assignment/Grading",
    "Tech Bug": "Technical Bug",
    "Certificte": "Certificate",
    "login": "Login/Access"
}

support["issue_category"] = support["issue_category"].replace(issue_category_map)


# 3. Standardize handled_by
handled_by_map = {
    "ai bot": "AI Bot",
    "AI-Bot": "AI Bot",
    "Bot": "AI Bot",
    "Human": "Human Agent",
    "human agent": "Human Agent"
}

support["handled_by"] = support["handled_by"].replace(handled_by_map)

In [296]:
for col in ["segment", "issue_category", "handled_by"]:
    print("\n", col)
    print(support[col].value_counts(dropna=False))


 segment
segment
K-12          1627
Test Prep     1190
Higher Ed     1096
Upskilling    1087
Name: count, dtype: int64

 issue_category
issue_category
Course Content Query    1118
Billing/Refund          1006
Assignment/Grading       886
Login/Access             861
Technical Bug            822
Certificate              307
Name: count, dtype: int64

 handled_by
handled_by
Human Agent        3507
AI Bot             1027
AI Bot -> Human     373
NaN                  93
Name: count, dtype: int64


In [319]:
support[[
    "first_response_mins",
    "resolution_hrs",
    "csat_score",
    "cost_usd"
]].describe()

,first_response_mins,resolution_hrs,csat_score,cost_usd
count,4901.000000,4901.000000,3468.000000,5000.000000
mean,376.494226,19.320461,3.865340,7.031424
std,4035.513178,85.344161,1.002465,19.092091
min,-5.000000,-2.000000,0.000000,-3.000000
25%,5.400000,4.260000,3.000000,4.407500
50%,161.100000,10.080000,4.000000,7.090000
75%,304.700000,18.390000,5.000000,8.820000
max,99999.000000,1500.000000,9.000000,450.000000


first_response_mins data-quality problems- both min and max are not logical
Both extremes are suspicious.
-5 minutes is impossible → invalid value.
99,999 minutes is extremely suspicious and is likely an erroneous/outlier value.
Notice how the mean (376) is much higher than the median (161). The extreme value is likely pulling the mean upward.

In [321]:
support[
    (support["first_response_mins"] < 0) |
    (support["first_response_mins"] > 1440)
][[
    "ticket_id",
    "first_response_mins",
    "resolution_hrs",
    "ai_phase"
]]

,ticket_id,first_response_mins,resolution_hrs,ai_phase
90,T502319,-5.0,32.88,AI Adoption
715,T501443,1523.6,10.85,Pre-AI
1373,T500401,-5.0,17.07,Pre-AI
1383,T502837,99999.0,0.15,AI Adoption
1690,T503350,-5.0,NaN,Post-AI
1717,T500137,-5.0,48.34,Pre-AI
1832,T504788,-5.0,12.53,Post-AI
1911,T500783,1817.6,25.23,Pre-AI
1970,T500218,-5.0,4.73,Pre-AI
2181,T501917,1449.6,25.28,Pre-AI


In [323]:
support[
    support["first_response_mins"].isin([-5, 99999])
][[
    "ticket_id",
    "first_response_mins",
    "resolution_hrs",
    "ai_phase"
]]

,ticket_id,first_response_mins,resolution_hrs,ai_phase
90,T502319,-5.0,32.88,AI Adoption
1373,T500401,-5.0,17.07,Pre-AI
1383,T502837,99999.0,0.15,AI Adoption
1690,T503350,-5.0,NaN,Post-AI
1717,T500137,-5.0,48.34,Pre-AI
1832,T504788,-5.0,12.53,Post-AI
1970,T500218,-5.0,4.73,Pre-AI
2289,T503164,-5.0,4.94,Post-AI
2427,T502675,-5.0,7.57,AI Adoption
2471,T501264,99999.0,5.15,Pre-AI


In [325]:
support["first_response_mins"] = support["first_response_mins"].replace(
    [-5, 99999],
    np.nan
)

In [327]:
support["first_response_mins"].describe()

count    4877.000000
mean      214.329752
std       235.561971
min         0.000000
25%         5.500000
50%       161.300000
75%       304.200000
max      1989.900000
Name: first_response_mins, dtype: float64

resolution_hrs data-quality problems- both min and max are not logical
-2 hours → impossible
1,500 hours → extremely large and should be investigated
Mean is almost twice the median → strong indication of a long right tail/outlier influence.

In [331]:
support[
    (support["resolution_hrs"] < 0) |
    (support["resolution_hrs"] > 720)
][[
    "ticket_id",
    "resolution_hrs",
    "first_response_mins",
    "ai_phase"
]]

,ticket_id,resolution_hrs,first_response_mins,ai_phase
26,T502549,1500.0,6.9,AI Adoption
217,T503686,-2.0,9.7,Post-AI
260,T504200,-2.0,8.1,Post-AI
441,T501834,1500.0,177.1,Pre-AI
467,T501135,-2.0,93.9,Pre-AI
647,T503413,1500.0,1.0,Post-AI
820,T502620,-2.0,755.4,AI Adoption
1035,T502748,1500.0,332.1,AI Adoption
1044,T502668,1500.0,304.8,AI Adoption
1226,T503025,1500.0,1.9,Post-AI


In [333]:
support["resolution_hrs"] = support["resolution_hrs"].replace(
    [-2, 1500],
    np.nan
)

In [335]:
support["resolution_hrs"].describe()

count    4878.000000
mean       15.726031
std        43.698221
min         0.050000
25%         4.310000
50%        10.075000
75%        18.340000
max       720.000000
Name: resolution_hrs, dtype: float64

Business-rule checks

In [344]:
support["reopened"].value_counts()

reopened
0    4516
1     484
Name: count, dtype: int64

In [346]:
support["escalated"].value_counts()

escalated
0    4197
1     803
Name: count, dtype: int64

#### Instructors Data

In [367]:
instructors.shape

(2293, 14)

In [369]:
instructors.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2293 entries, 0 to 2292
Data columns (total 14 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   instructor_id          2293 non-null   object 
 1   month                  2293 non-null   object 
 2   subject_area           2293 non-null   object 
 3   employment_type        2293 non-null   object 
 4   live_teaching_hrs      2293 non-null   float64
 5   doubt_resolution_hrs   2293 non-null   float64
 6   grading_hrs            2224 non-null   float64
 7   content_units_created  2225 non-null   float64
 8   content_creation_hrs   2293 non-null   float64
 9   ai_tool_usage_pct      2225 non-null   float64
 10  students_assigned      2224 non-null   float64
 11  avg_learner_rating     2225 non-null   float64
 12  instructor_cost_usd    2293 non-null   float64
 13  ai_phase               2293 non-null   object 
dtypes: float64(9), object(5)
memory usage: 250.9+ KB


1. Change the month data type

In [374]:
instructors["month"].unique()

array(['2025-08-01', '2024-10-01', '2023-09-01', '2023-06-01',
       '2025-04-01', '2022-05-01', '2022-03-01', '2022-07-01',
       '2022-09-01', '2026-05-01', '2025-01-01', '2026-07-01',
       '2023-05-01', '2026-03-01', '2024-05-01', '2024-03-01',
       '2026-02-01', '2025-11-01', '2021-11-01', '2024-06-01',
       '2024-02-01', '2024-11-01', '2023-12-01', '2023-08-01',
       '2024-07-01', '2025-12-01', '2025-06-01', '2021-04-01',
       '2023-07-01', '2022-01-01', '2022/07', '2025-05-01', '2023-11-01',
       '2023-10-01', '2022-12-01', '2026-04-01', '2022-11-01',
       '2026-08-01', '2024-12-01', '2026-01-01', '2021-12-01',
       '2023-01-01', '2024-04-01', '2024-08-01', '2023-02-01',
       '2021-10-01', '2023-04-01', '2022-10-01', 'May 2025', '2024-01-01',
       '2022-02-01', '2021-05-01', '2026-06-01', '2025-10-01', 'Mar-2025',
       '2022-08-01', '2025-09-01', '2025-03-01', '02/2025', '2021-03-01',
       '2021-06-01', '2025-02-01', 'February 2025', '2024-09-01',
      

In [376]:
instructors["month"] = pd.to_datetime(
    instructors["month"],
    format="mixed"
)

In [380]:
instructors.isna().sum()

instructor_id             0
month                     0
subject_area              0
employment_type           0
live_teaching_hrs         0
doubt_resolution_hrs      0
grading_hrs              69
content_units_created    68
content_creation_hrs      0
ai_tool_usage_pct        68
students_assigned        69
avg_learner_rating       68
instructor_cost_usd       0
ai_phase                  0
dtype: int64

In [384]:
instructors.duplicated().sum()

22

In [386]:
instructors.duplicated(subset=["instructor_id", "month"]).sum()

22

In [390]:
duplicates = instructors[
    instructors.duplicated(
        subset=["instructor_id", "month"],
        keep=False
    )
].sort_values(["instructor_id", "month"])

duplicates

,instructor_id,month,subject_area,employment_type,live_teaching_hrs,doubt_resolution_hrs,grading_hrs,content_units_created,content_creation_hrs,ai_tool_usage_pct,students_assigned,avg_learner_rating,instructor_cost_usd,ai_phase
122,I2001,2026-08-01,Programming,Freelance,15.95,4.80,3.71,0.0,0.00,58.77,93.0,4.52,865.17,Post-AI
1923,I2001,2026-08-01,Programming,Freelance,15.95,4.80,3.71,0.0,0.00,58.77,93.0,4.52,865.17,Post-AI
300,I2003,2026-05-01,Exam Prep (JEE/NEET/SAT),Part-time,35.13,8.42,7.95,3.0,21.15,57.39,173.0,4.29,2288.64,Post-AI
1101,I2003,2026-05-01,Exam Prep (JEE/NEET/SAT),Part-time,35.13,8.42,7.95,3.0,21.15,57.39,173.0,4.29,2288.64,Post-AI
1752,I2005,2021-07-01,Math,Part-time,30.83,14.15,8.49,7.0,55.51,0.00,144.0,4.15,2911.59,Pre-AI
2200,I2005,2021-07-01,Math,Part-time,30.83,14.15,8.49,7.0,55.51,0.00,144.0,4.15,2911.59,Pre-AI
871,I2005,2023-09-01,Math,Part-time,31.25,14.32,7.28,4.0,33.20,18.01,120.0,4.39,2392.57,AI Adoption
2160,I2005,2023-09-01,Math,Part-time,31.25,14.32,7.28,4.0,33.20,18.01,120.0,4.39,2392.57,AI Adoption
233,I2009,2021-08-01,Business & Finance,Full-time,60.03,34.31,23.09,8.0,68.42,0.00,290.0,4.38,6453.21,Pre-AI
1773,I2009,2021-08-01,Business & Finance,Full-time,60.03,34.31,23.09,8.0,68.42,0.00,290.0,4.38,6453.21,Pre-AI


In [392]:
instructors = instructors.drop_duplicates()

In [394]:
instructors.duplicated(subset=["instructor_id", "month"]).sum()

0

Invalid Ranges

In [397]:
instructors["ai_tool_usage_pct"].describe()

count    2203.000000
mean       39.700726
std        33.481076
min        -5.000000
25%         4.135000
50%        39.530000
75%        64.690000
max       130.000000
Name: ai_tool_usage_pct, dtype: float64

In [405]:
instructors[
    (instructors["ai_tool_usage_pct"] < 0) |
    (instructors["ai_tool_usage_pct"] > 100)
][[
    "instructor_id",
    "month",
    "ai_phase",
    "ai_tool_usage_pct"
]]

,instructor_id,month,ai_phase,ai_tool_usage_pct
55,I2033,2025-05-01,Post-AI,130.0
114,I2000,2025-04-01,Post-AI,-5.0
136,I2005,2024-11-01,Post-AI,-5.0
348,I2029,2025-03-01,Post-AI,130.0
607,I2013,2023-05-01,AI Adoption,-5.0
742,I2021,2022-01-01,Pre-AI,-5.0
751,I2040,2024-06-01,Post-AI,-5.0
791,I2008,2023-03-01,AI Adoption,-5.0
891,I2005,2025-12-01,Post-AI,130.0
907,I2024,2024-09-01,Post-AI,130.0


In [407]:
instructors.loc[
    (instructors["ai_tool_usage_pct"] < 0) |
    (instructors["ai_tool_usage_pct"] > 100),
    "ai_tool_usage_pct"
] = np.nan

In [409]:
instructors["ai_tool_usage_pct"].describe()

count    2181.000000
mean       39.532646
std        32.922095
min         0.000000
25%         4.910000
50%        39.540000
75%        64.480000
max       100.000000
Name: ai_tool_usage_pct, dtype: float64

In [399]:
instructors["avg_learner_rating"].describe()

count    2204.000000
mean        4.477037
std         2.471480
min         0.000000
25%         4.260000
50%         4.340000
75%         4.420000
max        48.000000
Name: avg_learner_rating, dtype: float64

In [411]:
instructors[
    (instructors["avg_learner_rating"] < 0) |
    (instructors["avg_learner_rating"] > 5)
][[
    "instructor_id",
    "month",
    "ai_phase",
    "avg_learner_rating"
]]

,instructor_id,month,ai_phase,avg_learner_rating
118,I2026,2023-02-01,Pre-AI,6.0
595,I2028,2022-05-01,Pre-AI,48.0
718,I2042,2022-10-01,Pre-AI,48.0
855,I2022,2023-08-01,AI Adoption,6.0
916,I2002,2024-04-01,Post-AI,6.0
985,I2005,2025-03-01,Post-AI,6.0
1018,I2037,2025-04-01,Post-AI,6.0
1256,I2012,2025-04-01,Post-AI,6.0
1261,I2026,2024-02-01,AI Adoption,48.0
1267,I2005,2024-10-01,Post-AI,48.0


In [413]:
instructors["avg_learner_rating"].sort_values().unique()

array([ 0.  ,  3.96,  3.97,  3.98,  3.99,  4.02,  4.04,  4.05,  4.06,
        4.07,  4.08,  4.09,  4.1 ,  4.11,  4.12,  4.13,  4.14,  4.15,
        4.16,  4.17,  4.18,  4.19,  4.2 ,  4.21,  4.22,  4.23,  4.24,
        4.25,  4.26,  4.27,  4.28,  4.29,  4.3 ,  4.31,  4.32,  4.33,
        4.34,  4.35,  4.36,  4.37,  4.38,  4.39,  4.4 ,  4.41,  4.42,
        4.43,  4.44,  4.45,  4.46,  4.47,  4.48,  4.49,  4.5 ,  4.51,
        4.52,  4.53,  4.54,  4.55,  4.56,  4.57,  4.58,  4.59,  4.6 ,
        4.61,  4.62,  4.63,  4.64,  4.65,  4.66,  4.67,  4.68,  4.69,
        4.7 ,  4.73,  4.74,  4.78,  4.88,  6.  , 48.  ,   nan])

In [415]:
instructors.loc[
    instructors["avg_learner_rating"] > 5,
    "avg_learner_rating"
] = np.nan

In [417]:
instructors["avg_learner_rating"].describe()

count    2187.000000
mean        4.330768
std         0.240723
min         0.000000
25%         4.260000
50%         4.340000
75%         4.420000
max         4.880000
Name: avg_learner_rating, dtype: float64

Negative hours

In [420]:
hour_cols = [
    "live_teaching_hrs",
    "doubt_resolution_hrs",
    "grading_hrs",
    "content_creation_hrs"
]

instructors[
    (instructors[hour_cols] < 0).any(axis=1)
][["instructor_id", "month"] + hour_cols]

,instructor_id,month,live_teaching_hrs,doubt_resolution_hrs,grading_hrs,content_creation_hrs
416,I2009,2024-06-01,-12.0,27.73,16.57,53.89
1099,I2034,2025-05-01,-12.0,24.86,16.98,34.40
1344,I2005,2025-05-01,-12.0,11.54,7.41,33.13
1487,I2012,2025-06-01,-12.0,5.59,4.24,19.44
1600,I2042,2022-04-01,-12.0,7.72,4.85,7.88
1638,I2013,2025-08-01,-12.0,23.93,15.79,44.46
2224,I2041,2023-08-01,-12.0,31.96,20.01,83.64


In [422]:
instructors.loc[
    instructors["live_teaching_hrs"] < 0,
    "live_teaching_hrs"
] = np.nan

In [424]:
(instructors["live_teaching_hrs"] < 0).sum()

0

In [428]:
count_cols = [
    "content_units_created",
    "students_assigned"
]

instructors[
    (instructors[count_cols] < 0).any(axis=1)
][["instructor_id", "month"] + count_cols]

,instructor_id,month,content_units_created,students_assigned


Suspicious zero values

In [430]:
instructors[
    (instructors["students_assigned"] == 0) &
    (instructors["live_teaching_hrs"] > 0)
]

,instructor_id,month,subject_area,employment_type,live_teaching_hrs,doubt_resolution_hrs,grading_hrs,content_units_created,content_creation_hrs,ai_tool_usage_pct,students_assigned,avg_learner_rating,instructor_cost_usd,ai_phase


In [432]:
instructors[
    (instructors["content_units_created"] == 0) &
    (instructors["content_creation_hrs"] > 0)
]

,instructor_id,month,subject_area,employment_type,live_teaching_hrs,doubt_resolution_hrs,grading_hrs,content_units_created,content_creation_hrs,ai_tool_usage_pct,students_assigned,avg_learner_rating,instructor_cost_usd,ai_phase


Categorical values

In [435]:
for col in ["subject_area", "employment_type", "ai_phase"]:
    print("\n", col)
    print(instructors[col].value_counts(dropna=False))


 subject_area
subject_area
Exam Prep (JEE/NEET/SAT)    537
Programming                 397
Data & AI                   326
Business & Finance          322
Language Learning           254
Math                        234
Science                     111
exam prep (jee/neet/sat)     24
programming                  19
data & ai                    15
math                         12
business & finance           10
language learning             8
science                       2
Name: count, dtype: int64

 employment_type
employment_type
Freelance     741
Full-time     727
Part-time     667
Freelancer     26
PT             24
FT             22
Full time      20
Part time      18
freelance      15
full-time      11
Name: count, dtype: int64

 ai_phase
ai_phase
Post-AI        1236
Pre-AI          520
AI Adoption     515
Name: count, dtype: int64


In [438]:
# Standardize subject_area
instructors["subject_area"] = instructors["subject_area"].replace({
    "exam prep (jee/neet/sat)": "Exam Prep (JEE/NEET/SAT)",
    "programming": "Programming",
    "data & ai": "Data & AI",
    "math": "Math",
    "business & finance": "Business & Finance",
    "language learning": "Language Learning",
    "science": "Science"
})


# Standardize employment_type
instructors["employment_type"] = instructors["employment_type"].replace({
    "Freelancer": "Freelance",
    "freelance": "Freelance",
    "FT": "Full-time",
    "Full time": "Full-time",
    "full-time": "Full-time",
    "PT": "Part-time",
    "Part time": "Part-time"
})

In [440]:
for col in ["subject_area", "employment_type"]:
    print("\n", col)
    print(instructors[col].value_counts())


 subject_area
subject_area
Exam Prep (JEE/NEET/SAT)    561
Programming                 416
Data & AI                   341
Business & Finance          332
Language Learning           262
Math                        246
Science                     113
Name: count, dtype: int64

 employment_type
employment_type
Freelance    782
Full-time    780
Part-time    709
Name: count, dtype: int64


Extreme/outlier values

In [442]:
instructors.describe().T

,count,mean,min,25%,50%,75%,max,std
month,2271,2024-04-11 17:43:59.365917952,2021-02-01 00:00:00,2023-04-01 00:00:00,2024-05-01 00:00:00,2025-06-01 00:00:00,2026-08-01 00:00:00,NaN
live_teaching_hrs,2264.0,36.898202,0.0,17.97,30.885,57.4525,310.0,24.404841
doubt_resolution_hrs,2271.0,15.812206,3.92,7.405,12.96,23.25,42.87,9.523009
grading_hrs,2203.0,10.491021,2.93,4.67,8.41,16.28,27.06,6.396345
content_units_created,2203.0,4.521562,0.0,2.0,4.0,6.0,21.0,3.052204
content_creation_hrs,2271.0,32.529956,0.0,16.36,28.11,45.545,134.45,21.686279
ai_tool_usage_pct,2181.0,39.532646,0.0,4.91,39.54,64.48,100.0,32.922095
students_assigned,2203.0,170.356786,51.0,77.0,135.0,267.0,425.0,104.099455
avg_learner_rating,2187.0,4.330768,0.0,4.26,4.34,4.42,4.88,0.240723
instructor_cost_usd,2271.0,3839.045711,-500.0,1677.87,2311.75,5513.505,250000.0,11748.819643


instructor_cost_usd has a serious issue
min = -500
max = 250000

In [445]:
instructors[
    (instructors["instructor_cost_usd"] < 0) |
    (instructors["instructor_cost_usd"] > 20000)
][[
    "instructor_id",
    "month",
    "employment_type",
    "ai_phase",
    "instructor_cost_usd"
]]

,instructor_id,month,employment_type,ai_phase,instructor_cost_usd
266,I2020,2021-10-01,Part-time,Pre-AI,-500.0
473,I2004,2023-09-01,Full-time,AI Adoption,250000.0
626,I2036,2024-12-01,Freelance,Post-AI,250000.0
840,I2000,2024-05-01,Freelance,Post-AI,250000.0
878,I2028,2022-07-01,Part-time,Pre-AI,250000.0
1110,I2009,2025-11-01,Full-time,Post-AI,-500.0
1368,I2018,2023-04-01,Freelance,AI Adoption,250000.0
1565,I2030,2021-12-01,Part-time,Pre-AI,-500.0
1959,I2030,2023-08-01,Part-time,AI Adoption,-500.0


In [447]:
instructors.loc[
    (instructors["instructor_cost_usd"] < 0) |
    (instructors["instructor_cost_usd"] > 20000),
    "instructor_cost_usd"
] = np.nan

In [449]:
instructors["instructor_cost_usd"].describe()

count    2262.000000
mean     3302.596291
std      2064.481530
min         0.000000
25%      1680.330000
50%      2311.405000
75%      5504.415000
max      9324.560000
Name: instructor_cost_usd, dtype: float64

#### Learners Data

In [460]:
learners.shape

(5000, 26)

In [462]:
learners.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5000 entries, 0 to 4999
Data columns (total 26 columns):
 #   Column                     Non-Null Count  Dtype  
---  ------                     --------------  -----  
 0   record_id                  5000 non-null   object 
 1   learner_id                 5000 non-null   object 
 2   signup_date                5000 non-null   object 
 3   country                    4950 non-null   object 
 4   segment                    5000 non-null   object 
 5   course_category            5000 non-null   object 
 6   plan_type                  5000 non-null   object 
 7   acquisition_channel        4901 non-null   object 
 8   ai_feature_used            2288 non-null   object 
 9   ai_sessions_per_week       4901 non-null   float64
 10  avg_session_min            4850 non-null   float64
 11  human_tutor_hrs_per_month  4851 non-null   float64
 12  courses_enrolled           5000 non-null   int64  
 13  completion_rate_pct        4807 non-null   float

1. signup_date , last_active_date and monthly_fee_usd data type should be change

In [471]:
learners["last_active_date"] = pd.to_datetime(
    learners["last_active_date"],
    format="mixed"
)

In [473]:
learners["signup_date"] = pd.to_datetime(
    learners["signup_date"],
    format="mixed"
)

In [477]:
learners["monthly_fee_usd"].unique()

array(['9.0', '19.0', '$0.0', '0.0', '49.0', '4067.0', '$9.0', '1577.0',
       '747.0', '$19.0', '19', '$49.0'], dtype=object)

In [480]:
learners["monthly_fee_usd"].value_counts()

monthly_fee_usd
9.0       1598
19.0      1328
0.0       1178
49.0       603
$9.0        61
$19.0       59
$0.0        50
1577.0      32
$49.0       27
747.0       26
19          26
4067.0      12
Name: count, dtype: int64

In [ ]:
learners["monthly_fee_usd"] = (
    learners["monthly_fee_usd"]
    .str.replace("$", "", regex=False)
    .astype(float)
)

In [484]:
learners["monthly_fee_usd"].value_counts().sort_index()

monthly_fee_usd
0.0       1228
9.0       1659
19.0      1413
49.0       630
747.0       26
1577.0      32
4067.0      12
Name: count, dtype: int64

In [490]:
learners["monthly_fee_usd"].dtype

dtype('float64')

In [492]:
learners.duplicated().sum()

96

In [494]:
learners["record_id"].duplicated().sum()

96

In [496]:
learners["learner_id"].duplicated().sum()

272

In [502]:
learners["learner_id"].value_counts().value_counts().sort_index()

count
1    4468
2     248
3      12
Name: count, dtype: int64

In [504]:
learners[learners["learner_id"].duplicated(keep=False)][
    ["learner_id", "record_id", "course_category", "signup_date"]
].sort_values("learner_id").head(30)

,learner_id,record_id,course_category,signup_date
1872,L10040,R101130,Math,2023-04-22
1047,L10040,R200008,Math,2023-04-22
4731,L10677,R100582,Exam Prep (JEE/NEET/SAT),2021-08-09
3802,L10677,R102182,Business & Finance,2024-06-07
1664,L10739,R100186,Exam Prep (JEE/NEET/SAT),2021-12-10
3118,L10739,R200049,Exam Prep (JEE/NEET/SAT),2021-12-10
634,L11221,R100966,Data & AI,2024-05-29
44,L11221,R200018,Data & AI,2024-05-29
1089,L11313,R104289,Science,2022-09-30
2412,L11313,R104286,Language Learning,2025-01-17


In [506]:
learners[learners.duplicated(keep=False)].shape

(192, 26)

In [510]:
learners = learners.drop_duplicates()

In [512]:
learners[learners.duplicated(keep=False)].shape

(0, 26)

Invalid categorical values

In [523]:
learners["churned"].value_counts(dropna=False)

churned
0        3222
1        1442
FALSE      65
N          61
No         42
Y          30
Yes        28
TRUE       14
Name: count, dtype: int64

In [527]:
learners["churned"] = (
    learners["churned"]
    .astype(str)
    .str.strip()
    .str.upper()
    .map({
        "0": 0,
        "FALSE": 0,
        "N": 0,
        "NO": 0,
        "1": 1,
        "TRUE": 1,
        "Y": 1,
        "YES": 1
    })
)

In [529]:
learners["churned"].value_counts(dropna=False)

churned
0    3390
1    1514
Name: count, dtype: int64

In [531]:
learners["course_category"].value_counts(dropna=False)

course_category
Data & AI                   738
Programming                 720
Science                     711
Business & Finance          702
Exam Prep (JEE/NEET/SAT)    687
Math                        677
Language Learning           669
Name: count, dtype: int64

In [535]:
learners["plan_type"].value_counts(dropna=False)

plan_type
Basic             1534
Premium           1298
Free              1145
Premium+Tutor      586
basic               89
Premuim             73
Prem+Tutor          69
free                27
 Basic              23
premium             18
 Premium            17
 Premium+Tutor      11
 Free                9
premium+tutor        5
Name: count, dtype: int64

In [537]:
learners["plan_type"] = (
    learners["plan_type"]
    .astype(str)
    .str.strip()
    .str.lower()
    .replace({
        "basic": "Basic",
        "premium": "Premium",
        "premuim": "Premium",
        "free": "Free",
        "premium+tutor": "Premium+Tutor",
        "prem+tutor": "Premium+Tutor"
    })
)

In [539]:
learners["plan_type"].value_counts(dropna=False)

plan_type
Basic            1646
Premium          1406
Free             1181
Premium+Tutor     671
Name: count, dtype: int64

In [541]:
learners["acquisition_channel"].value_counts(dropna=False)

acquisition_channel
Paid Social       1204
Organic Search    1204
YouTube            715
Referral           707
Partnerships       501
Email              477
NaN                 96
Name: count, dtype: int64

In [543]:
learners["ai_feature_used"].value_counts(dropna=False)

ai_feature_used
NaN                   2653
AI Doubt Solver        664
AI Tutor               597
AI Quiz Generator      339
AI Essay Feedback      280
AI Support Chatbot     234
AI tutor                38
ai doubt solver         34
AI Doubt-Solver         34
none                    31
Name: count, dtype: int64

In [545]:
learners["ai_feature_used"] = (
    learners["ai_feature_used"]
    .astype("string")
    .str.strip()
    .str.lower()
    .replace({
        "ai tutor": "AI Tutor",
        "ai doubt solver": "AI Doubt Solver",
        "ai doubt-solver": "AI Doubt Solver",
        "ai quiz generator": "AI Quiz Generator",
        "ai essay feedback": "AI Essay Feedback",
        "ai support chatbot": "AI Support Chatbot",
        "none": "None"
    })
)

In [547]:
learners["ai_feature_used"].value_counts(dropna=False)

ai_feature_used
<NA>                  2653
AI Doubt Solver        732
AI Tutor               635
AI Quiz Generator      339
AI Essay Feedback      280
AI Support Chatbot     234
None                    31
Name: count, dtype: Int64

In [553]:
learners["segment"].value_counts(dropna=False)

segment
K-12          1484
Test Prep     1142
Higher Ed     1006
Upskilling     993
Higher-Ed       68
K12             64
Test prep       61
k-12            21
HIGHER ED       15
test prep       11
upskilling      11
TEST PREP       10
UPSKILLING      10
higher ed        8
Name: count, dtype: int64

In [559]:
learners["segment"] = (
    learners["segment"]
    .astype("string")
    .str.strip()
    .str.lower()
    .replace({
        "k-12": "K-12",
        "k12": "K-12",
        "test prep": "Test Prep",
        "higher ed": "Higher Ed",
        "higher-ed": "Higher Ed",
        "upskilling": "Upskilling"
    })
)

In [561]:
learners["segment"].value_counts(dropna=False)

segment
K-12          1569
Test Prep     1224
Higher Ed     1097
Upskilling    1014
Name: count, dtype: Int64

In [563]:
learners["country"].value_counts(dropna=False)

country
India             1655
United States      877
United Kingdom     365
Brazil             317
Philippines        301
Germany            273
UAE                248
Nigeria            233
india               63
Bharat              63
IN                  61
INDIA               50
NaN                 49
BR                  35
United States       31
USA                 30
U.S.A               29
Brasil              25
nigeria             23
Dubai               22
UK                  20
U.A.E               18
PH                  18
us                  17
DE                  16
united kingdom      14
NG                  14
Deutschland         13
Philipines          13
U.K.                11
Name: count, dtype: int64

In [565]:
learners["country"] = (
    learners["country"]
    .astype("string")
    .str.strip()
    .str.lower()
    .replace({
        "india": "India",
        "bharat": "India",
        "in": "India",

        "united states": "United States",
        "usa": "United States",
        "u.s.a": "United States",
        "us": "United States",

        "united kingdom": "United Kingdom",
        "uk": "United Kingdom",
        "u.k.": "United Kingdom",

        "brazil": "Brazil",
        "brasil": "Brazil",
        "br": "Brazil",

        "philippines": "Philippines",
        "ph": "Philippines",
        "philipines": "Philippines",

        "germany": "Germany",
        "de": "Germany",
        "deutschland": "Germany",

        "uae": "UAE",
        "u.a.e": "UAE",
        "dubai": "UAE",

        "nigeria": "Nigeria",
        "ng": "Nigeria"
    })
)

In [567]:
learners["country"].value_counts(dropna=False)

country
India             1892
United States      984
United Kingdom     410
Brazil             377
Philippines        332
Germany            302
UAE                288
Nigeria            270
<NA>                49
Name: count, dtype: Int64

In [575]:
learners.describe()

,signup_date,ai_sessions_per_week,avg_session_min,human_tutor_hrs_per_month,courses_enrolled,completion_rate_pct,avg_quiz_score,support_tickets_90d,avg_ticket_resolution_hrs,csat_score,nps_score,monthly_fee_usd,ai_inference_cost_usd,cost_to_serve_usd,refund_flag,churned,last_active_date,tenure_months
count,4904,4805.000000,4756.000000,4757.000000,4904.000000,4713.000000,4706.000000,4764.000000,4661.000000,4515.000000,4565.000000,4904.000000,4904.000000,4805.000000,4904.000000,4904.000000,4904,4807.000000
mean,2023-10-26 08:08:54.420880896,1.643913,33.531960,1.003069,3.518352,60.941969,66.006715,1.989295,11.346385,3.830786,5.524863,38.683320,0.415844,11.490966,0.027936,0.308728,2026-03-03 22:55:06.362153472,29.316830
min,2021-01-02 00:00:00,0.000000,-15.000000,0.000000,1.000000,-5.000000,20.400000,-1.000000,0.500000,0.000000,0.000000,0.000000,0.000000,-3.200000,0.000000,0.000000,2020-12-31 00:00:00,0.000000
25%,2022-05-24 18:00:00,0.000000,24.000000,0.000000,2.000000,46.800000,56.500000,1.000000,7.100000,3.000000,4.000000,9.000000,0.000000,4.180000,0.000000,0.000000,2026-02-23 00:00:00,14.000000
50%,2023-10-16 12:00:00,0.000000,31.000000,0.000000,4.000000,61.100000,66.000000,2.000000,11.400000,4.000000,6.000000,9.000000,0.000000,6.100000,0.000000,0.000000,2026-08-24 00:00:00,27.000000
75%,2025-04-01 06:00:00,2.800000,37.000000,1.300000,5.000000,75.600000,75.700000,3.000000,15.500000,4.000000,7.000000,19.000000,0.650000,8.900000,0.000000,1.000000,2026-08-31 00:00:00,44.000000
max,2026-10-04 00:00:00,16.500000,999.000000,10.500000,6.000000,150.000000,100.000000,60.000000,29.700000,11.000000,10.000000,4067.000000,5.370000,999.990000,1.000000,1.000000,2026-11-05 00:00:00,68.000000
std,NaN,2.372582,52.535678,1.864868,1.706268,21.411779,14.145116,3.454405,5.583597,1.043655,2.433985,242.084073,0.652214,49.949958,0.164807,0.462016,NaN,18.865153


Numeric range

In [580]:
learners["completion_rate_pct"].describe()

count    4713.000000
mean       60.941969
std        21.411779
min        -5.000000
25%        46.800000
50%        61.100000
75%        75.600000
max       150.000000
Name: completion_rate_pct, dtype: float64

In [588]:
learners[
    (learners["completion_rate_pct"] < 0) |
    (learners["completion_rate_pct"] > 100)
][["learner_id", "completion_rate_pct"]]

,learner_id,completion_rate_pct
355,L30413,110.0
396,L82332,150.0
488,L77707,104.5
549,L19019,110.0
626,L55108,110.0
651,L20715,-5.0
749,L68046,150.0
930,L22084,110.0
972,L93133,-5.0
1040,L82613,110.0


In [590]:
learners[
    (learners["completion_rate_pct"] < 0) |
    (learners["completion_rate_pct"] > 100)
][[
    "learner_id",
    "completion_rate_pct",
    "courses_enrolled",
    "course_category",
    "plan_type",
    "ai_feature_used",
    "churned"
]]

,learner_id,completion_rate_pct,courses_enrolled,course_category,plan_type,ai_feature_used,churned
355,L30413,110.0,1,Business & Finance,Premium+Tutor,<NA>,0
396,L82332,150.0,1,Programming,Basic,AI Essay Feedback,0
488,L77707,104.5,4,Exam Prep (JEE/NEET/SAT),Free,<NA>,1
549,L19019,110.0,5,Programming,Basic,<NA>,1
626,L55108,110.0,6,Math,Premium+Tutor,<NA>,1
651,L20715,-5.0,2,Programming,Premium,AI Tutor,0
749,L68046,150.0,3,Math,Free,AI Doubt Solver,0
930,L22084,110.0,1,Math,Basic,<NA>,0
972,L93133,-5.0,1,Exam Prep (JEE/NEET/SAT),Free,<NA>,0
1040,L82613,110.0,5,Business & Finance,Premium,AI Doubt Solver,0


In [592]:
learners.loc[
    (learners["completion_rate_pct"] < 0) |
    (learners["completion_rate_pct"] > 100),
    "completion_rate_pct"
] = pd.NA

In [582]:
learners["avg_session_min"].describe()

count    4756.000000
mean       33.531960
std        52.535678
min       -15.000000
25%        24.000000
50%        31.000000
75%        37.000000
max       999.000000
Name: avg_session_min, dtype: float64

In [594]:
learners[
    (learners["avg_session_min"] < 0) |
    (learners["avg_session_min"] > 120)
][[
    "learner_id",
    "avg_session_min",
    "courses_enrolled",
    "course_category",
    "plan_type",
    "ai_feature_used"
]]

,learner_id,avg_session_min,courses_enrolled,course_category,plan_type,ai_feature_used
12,L49609,-15.0,5,Business & Finance,Premium,AI Quiz Generator
484,L60452,-15.0,6,Business & Finance,Premium+Tutor,AI Tutor
576,L63761,-15.0,1,Data & AI,Basic,AI Support Chatbot
649,L61633,720.0,3,Science,Premium+Tutor,AI Doubt Solver
782,L21538,999.0,4,Exam Prep (JEE/NEET/SAT),Basic,<NA>
823,L43592,-15.0,2,Language Learning,Basic,<NA>
919,L21964,999.0,4,Exam Prep (JEE/NEET/SAT),Premium,AI Doubt Solver
1044,L65614,-15.0,4,Science,Premium+Tutor,<NA>
1171,L71840,999.0,3,Exam Prep (JEE/NEET/SAT),Premium,<NA>
1214,L36623,720.0,1,Data & AI,Free,<NA>


In [596]:
learners.loc[
    learners["avg_session_min"].isin([-15, 720, 999]),
    "avg_session_min"
] = pd.NA

In [598]:
learners["avg_session_min"].describe()

count    4727.000000
mean       30.499683
std         9.473148
min         0.000000
25%        24.000000
50%        30.000000
75%        37.000000
max        65.000000
Name: avg_session_min, dtype: float64

In [584]:
learners["support_tickets_90d"].describe()

count    4764.000000
mean        1.989295
std         3.454405
min        -1.000000
25%         1.000000
50%         2.000000
75%         3.000000
max        60.000000
Name: support_tickets_90d, dtype: float64

In [600]:
learners[
    learners["support_tickets_90d"] < 0
][[
    "learner_id",
    "support_tickets_90d",
    "plan_type",
    "course_category",
    "ai_feature_used",
    "churned"
]]

,learner_id,support_tickets_90d,plan_type,course_category,ai_feature_used,churned
376,L75163,-1.0,Premium,Language Learning,AI Tutor,0
444,L20073,-1.0,Basic,Language Learning,<NA>,0
800,L33174,-1.0,Premium+Tutor,Programming,AI Quiz Generator,0
1083,L14189,-1.0,Premium,Language Learning,<NA>,1
2366,L53503,-1.0,Basic,Exam Prep (JEE/NEET/SAT),AI Quiz Generator,0
2781,L44125,-1.0,Basic,Data & AI,<NA>,1
3016,L30494,-1.0,Basic,Programming,<NA>,0
3150,L68243,-1.0,Premium,Language Learning,AI Quiz Generator,0
3469,L98411,-1.0,Free,Exam Prep (JEE/NEET/SAT),<NA>,1
3540,L20604,-1.0,Premium,Programming,<NA>,1


In [602]:
(learners["support_tickets_90d"] < 0).sum()

14

In [604]:
learners.loc[
    learners["support_tickets_90d"] < 0,
    "support_tickets_90d"
] = pd.NA

In [586]:
learners["cost_to_serve_usd"].describe()

count    4805.000000
mean       11.490966
std        49.949958
min        -3.200000
25%         4.180000
50%         6.100000
75%         8.900000
max       999.990000
Name: cost_to_serve_usd, dtype: float64

In [606]:
learners[
    (learners["cost_to_serve_usd"] < 0) |
    (learners["cost_to_serve_usd"] > 100)
][[
    "learner_id",
    "cost_to_serve_usd",
    "plan_type",
    "course_category",
    "ai_feature_used",
    "ai_inference_cost_usd"
]]

,learner_id,cost_to_serve_usd,plan_type,course_category,ai_feature_used,ai_inference_cost_usd
431,L14108,999.99,Basic,Math,<NA>,0.00
555,L58150,999.99,Basic,Exam Prep (JEE/NEET/SAT),AI Essay Feedback,0.09
638,L18002,999.99,Premium+Tutor,Business & Finance,AI Doubt Solver,0.00
837,L79483,999.99,Basic,Science,<NA>,0.00
839,L44346,450.00,Premium,Exam Prep (JEE/NEET/SAT),AI Doubt Solver,0.32
878,L72044,999.99,Basic,Language Learning,<NA>,0.00
893,L21712,-3.20,Basic,Data & AI,AI Support Chatbot,0.91
1113,L17594,450.00,Premium,Business & Finance,<NA>,0.00
1285,L71298,450.00,Free,Science,<NA>,0.00
1994,L55226,450.00,Premium+Tutor,Science,<NA>,0.00


In [608]:
learners.loc[
    learners["cost_to_serve_usd"].isin([-3.20, 450.00, 999.99]),
    "cost_to_serve_usd"
] = pd.NA

In [610]:
learners["cost_to_serve_usd"].isin([-3.20, 450.00, 999.99]).sum()

0

In [612]:
learners["cost_to_serve_usd"].describe()

count    4779.000000
mean        8.523413
std         7.210091
min         0.800000
25%         4.180000
50%         6.100000
75%         8.900000
max        33.600000
Name: cost_to_serve_usd, dtype: float64

In [616]:
#### A learner cannot have last_active_date before signup_date.
learners["last_active_date"] < learners["signup_date"]

0       False
1       False
2       False
3       False
4       False
        ...  
4995    False
4996    False
4997    False
4998    False
4999    False
Length: 4904, dtype: bool

In [618]:
learners[
    (learners["ai_feature_used"] == "No") &
    (learners["ai_sessions_per_week"] > 0)
]

,record_id,learner_id,signup_date,country,segment,course_category,plan_type,acquisition_channel,ai_feature_used,ai_sessions_per_week,...,avg_ticket_resolution_hrs,csat_score,nps_score,monthly_fee_usd,ai_inference_cost_usd,cost_to_serve_usd,refund_flag,churned,last_active_date,tenure_months
